pinot-core/src/test/java/com/linkedin/pinot/core/periodictask/PeriodicTaskSchedulerTest.java

```
@@ -0,0 +1,181 @@
+    long runFrequencyInSeconds = 1L;
+    long totalRunTimeInMilliseconds = 4_000L;
+
+    List<PeriodicTask> periodicTasks = new ArrayList<>();
+    PeriodicTask task = new BasePeriodicTask("Task", runFrequencyInSeconds, 0L) {
+      @Override
+      public void run() {
+        // Execute task.
+        count.incrementAndGet();
+      }
+    };
+    periodicTasks.add(task);
+
+    long start = System.currentTimeMillis();
+    periodicTaskScheduler.start(periodicTasks);
+    try {
+      Thread.sleep(totalRunTimeInMilliseconds);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+    periodicTaskScheduler.stop();
+
+    Assert.assertTrue(count.get() > 0);
+    Assert.assertTrue(count.get() < (totalRunTimeInMilliseconds / runFrequencyInSeconds));
+    Assert.assertTrue(totalRunTimeInMilliseconds < (System.currentTimeMillis() - start));
+  }
+
+  @Test
+  public void testSchedulerWithTwoStaggeredTasks() {
+    AtomicInteger count = new AtomicInteger(0);
+    PeriodicTaskScheduler periodicTaskScheduler = new PeriodicTaskScheduler(0L);
+    long runFrequencyInSeconds = 4L;
+    long totalRunTimeInMilliseconds = 15_000L;
+
+    List<PeriodicTask> periodicTasks = new ArrayList<>();
+    PeriodicTask task1 = new BasePeriodicTask("Task1", runFrequencyInSeconds, 0L) {
+      @Override
+      public void run() {
+        // Execute task.
+        count.incrementAndGet();
+      }
+    };
+    periodicTasks.add(task1);
+
+    // Stagger 2 tasks.
+    PeriodicTask task2 = new BasePeriodicTask("Task2", runFrequencyInSeconds, runFrequencyInSeconds / 2) {
+      @Override
+      public void run() {
+        // Execute task.
+        count.decrementAndGet();
+      }
+    };
+    periodicTasks.add(task2);
+
+    long start = System.currentTimeMillis();
+    periodicTaskScheduler.start(periodicTasks);
+    try {
+      Thread.sleep(totalRunTimeInMilliseconds);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
```

I am not sure catching interrupted exception is a good idea. You can add throws to the method signature and let the test fail and bail instead of continue on.